### PR TITLE
refactor: Migrate parameter configs to YAML

### PR DESCRIPTION
**Pull request by akashin**
This is a refactoring PR to enable https://github.com/near/nearcore/issues/8264. For full motivation see https://github.com/near/nearcore/issues/8264#issuecomment-1373628858

This PR only changes the disk format and parsing logic but doesn't change the in-memory representation, so it should not lead to any visible behaviour changes.

The next step in this refactoring would be to investigate getting rid of `serde_json::Value` in favor of `serde_yaml::Value` or even better a specialized struct with much more restricted types to catch invalid config errors earlier.

### CODE REVIEW

**Comment by akashin** [comment]
The build is currently fails because we use two versions of `serde_yaml`, one introduced in this PR: "0.9.16" and the other "0.8.24" used by `paperclip` and `insta` deps. I've tried to use 0.8.24 in this PR and it exposes two non-critical bugs in older YAML implementation that are caught by our tests.

We can either use 0.8.24 in this PR and fix the failing tests or update `paperclip` and `insta`, as both use `serde_yaml` "0.9.*" in their latest releases. I'm not sure what is the usual approach, so need some advice here.

**Comment by akashin** [comment]
I've tried to go the update-way, but ran into some problems described in https://github.com/near/nearcore/issues/8320 .
I will continue working on the update, but given that it is blocked on external parties, for now I will implement workarounds for this code to work with older YAML version and add TODOs to remove them after version upgrade.

**Review by jakmeier — APPROVED** [review]
@akashin Looks good to me, please address the one little nitpick and afterwards this can be merged as far as I am concerned. But let's settle the debate around the format first. :)

> Almost no change to existing config to turn it into YAML

That's kind of nice but the more I think about it, the more I am convinced this should not influence our decision too much. Long-term it doesn't matter how big the diff in the PR is.

> Overall, I think the plan is to keep the config relatively flat (up to 3 layers deep) and don't use any fancy YAML features (tags, references), so we hopefully avoid any portability and debugging issues. Also, we already depend on serde_yaml indirectly, so this is not as bad as introducing a new dependency.

I agree with everything here. Although, if we could remove the dependency on `serde_yaml` entirely, that would be nice and we are moving one step further away from that goal.

I guess I have some personal hatred towards yaml as the spec is rather complex and often tools are not fully compatible as a result of it. (At least it used to be like this and your struggle with `serde_yaml` sort of confirms that.^^) But as long as we keep it simple, as you suggested, this should not be an issue.

@nagisa do you want to veto against YAML? I think TOML would work fine here, too. I don't have a strong opinion in either direction. Both should allow to add a `weight` (new and old) to each parameter without too much hassle. But I can see the arguments presented by @akashin that it's a bit nicer in YAML.

**Comment by nagisa** [comment]
Not going to veto yaml. I had my fair share of terrible experiences with it, but mostly in the context of a project attempting to shoehorn it into an application where a DSL would have been more appropriate (ansible, github actions…) I heard horror stories about configuration files written in yaml too, but those also approaching CI definition levels of complexity.

File: core/primitives/src/runtime/parameter_table.rs [hunk 1]
```diff
@@ -206,58 +219,45 @@ impl ParameterTable {
     }
 }
 
+/// Represents YAML values supported by parameter diff config.
+#[derive(Deserialize, Clone, Debug)]
```
[the remark this answers is not in the document]
Comment (akashin) [reply core/primitives/src/runtime/parameter_table.rs]: Makes sense, done.

File: core/primitives/src/runtime/parameter_table.rs [hunk 2]
```diff
@@ -206,58 +219,45 @@ impl ParameterTable {
...
+
+                let old_value = if let Some(s) = &value.old {
+                    parse_parameter_txt_value(s)?
                 } else {
-                    Ok((
-                        typed_key,
-                        (serde_json::Value::Null, parse_parameter_txt_value(value.trim())?),
-                    ))
-                }
+                    serde_json::Value::Null
```
[the remark this answers is not in the document]
Comment (akashin) [reply core/primitives/src/runtime/parameter_table.rs:244]: Yes, I was hoping to get rid of this in the next refactoring PR, will give it a try and see how big of a change it is going to be. If that turns out too big, I'll add a TODO here.

**Comment by akashin** [comment]
Ok, given that there are no strong arguments against, let's try using YAML for now and reevaluate whether we need to switch to TOML/RON when we either:
- Run into YAML parsing/debugging issues
- Have problems evolving the config to support our needs

**Pull request merged**